### DISCUSSION TITLE
gcc incompatibility

### DISCUSSION THREAD

**Discussion by JeffBezsos** [linux-surface/linux-surface#1633] [Q&A]
boys n girls, sirs & ladies... beautiful people of this fine project! 

here's an issue I experience & as a noob I will try to describe the issue as well as I can but if it all sounds super nooby to you please don't get offended - I'm learning ;) 

I'm on Surface Book 2 with discreet GPU (NVIDIA GTX 1050) & running Pop_OS! 

2 days ago  a new kernel build has been introduced! 12.3 - amazing - I love it when things move forward, kudos to all who worked on this! I applaud you for the level of skill that I envy! 

I updated & upgraded & at 1st there were login issues - that was fixed in mere hours! - amazing job! 

what is still not fixed & breaks the usability of the current kernel - is the nvidia driver compatibility with the new build - as I check in the make.log files (after doing some digging & spending countless hours trying to fix this issue) **the new kernel version has been built using quite an old gcc version 9.4... used by Ubuntu 20.04** & here's where the issues start...  mind you this may be a unique thing that applies only to Pop_OS! users who rely on system76-nvidia drivers - which use a more up to date gcc version of 11.4... & hence the dGPU simply won't work... I tried to install the gcc 9.4... & use CC=... command to specify installation of the nvidia drivers... running via the matching gcc-9 to no avail... = still the same errors... 

my next step I'll try to fix this issue is to try to recompile the kernel with a matching gcc version to the one my system uses = 11.4 - & here's the question! 

1. will this work? - theoretically - I'm asking the more experienced in the linux dependencies pains users - from your experience does this sound like the right solution? 
2. how do I go about it? :D 

thank you for your attention I hope I explain this well enough for anyone to understand & if I didn't please guide me how!

**Comment by sean-skt** [comment]
I'd report this in "Issues". I'm having this problem too with 2x Surface Book 2's. One Surface Book 2 has Pop!_OS 22.04 and the other has Ubuntu 24.04. Same problem with both. Since updating both to 6.12.3-2 the nVidia GTX 1050's no longer show up in the nVidia driver control panel because of the very reason you mention. Why was this kernel built with such an old GCC version?

**Reply by JeffBezsos** [reply]
@sean-skt the more I've been digging into this the more I realize this is a pretty common issue in Linux when a new kernel version is introduced nvidia takes it's sweet time to adjust... ----> take a look through this comments section here on official nvidia website  ---> https://forums.developer.nvidia.com/t/linux-new-kernel-6-5-0-14-ubuntu-22-04-can-not-compile-nvidia-display-card-driver/278553   & also here ----> https://www.linux.org/threads/solved-nouveau-vs-proprietary-nvidia-drivers-performance.43668/  ---> you'll learn this a common reoccurring issue! = essentially you have a hardware from 1 producer (in this case microsoft), that relies on components (like your GPU, CPU...) from various vendors & your OS is made by 1 group of developers & requires projects like this one to patch up the mainstream OS & than this also means all sorts of 3rd party drivers (thank god nvidia open sources their drivers this year...)... etc... you see why people use macs? lol! 

so all we can do is wait until nvidia stands up to it & updates the drivers... 

if gcc was that big of a deal when you think of it - there would be compatibility issues way more often than experienced! 
the older gcc version choice must be due to legacy hardware & software - some folks here are running older surface devices than our book 2's... ----> but check this article if you want to play a lil bit... https://medium.com/@srikarthikv/installing-nvidia-drivers-when-the-normal-way-doesnt-work-a-deep-dive-into-gcc-kernel-versions-11c1434bb0ef    (1 caveat - if you run gcc-9 command it will most likely install 9.5 - which is a newer version that was used in ubuntu 22.04 so you need to do some digging on ubuntu repositories for the specific 9.4... one - just check the make.log in either the 12.3 kernel directory or the nvidia one - it will tell you they don't match - it's like the 2nd or 3rd line in the log... it will also tell you what specific version you need - the name must be an exact match...)  

what you can also do though while waiting for the nvidia magic to happen is to temporarily downgrade to the last kernel version that worked (if your life/work depends on the dGPU) & just wait a lil longer till nvidia catches up... (smthn to do with kernel headers & dkms...) - just set some conditions (hold) for the old kernel so that every time you run sudo apt update & upgrade commands the kernel won't update (until you unhold it) 

but apparently you can compile the nvidia drivers yourself... someone mentioned it worked for them in the past I have no time nor patience to do that I go for the easiest solution = temporarily downgrade the kernel & wait till it's fixed n polished... 

hope this is helpful! 

good luck!

**Reply by qzed** [reply]
The problem is not necessarily nvidia themselves, but the distribution. Ubuntu/Debian/Pop generally use an older stable kernel and base all their packages around that version. So all the drivers and driver packages they provide are designed specifically for those versions and if you now install a newer kernel they may be incompatible.

And unfortunately, for us it's not really feasible to maintain an older kernel version that more closely matches their version.

**Reply by JeffBezsos** [reply]
@qzed so you guys run arch based distros? - I use pop because of COSMIC (best DE I tested & now it's going mainstream so I might consider switching as as you pointed out pop/debian... are slow with updates...) COSMIC is best for my workflow & nvidia drivers work(ed) out of the box... the distro requires little to no maintenance (if it wasn't for the pesky surface firmware issues... damn you microsoft!) & i'll take any QOL over having to learn smthn I may encounter issues with once or twice a year... 

what distro do you recommend? (I'm too nooby for vanilla arch mind you :D)

**Reply by qzed** [reply]
I'm using Fedora right now. Used to use Arch. Both should work pretty well and I can recommend Fedora if you want something that works (mostly) out of the box. For nvidia drivers you'll have to enable rpmfusion, but that's fairly easy.

**Reply by Thevandermeij** [reply]
same problem

//$ sudo apt install linux-image-surface linux-headers-surface libwacom-surface iptsd surface-control surface-dtx-daemon
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
linux-headers-surface is already the newest version (6.13.6-surface-1).
libwacom-surface is already the newest version (2.13.0-2).
iptsd is already the newest version (3-1).
surface-control is already the newest version (0.4.7-1).
surface-dtx-daemon is already the newest version (0.3.8-1).
The following additional packages will be installed:
  linux-image-6.13.6-surface-1
The following NEW packages will be installed:
  linux-image-6.13.6-surface-1 linux-image-surface
0 upgraded, 2 newly installed, 0 to remove and 0 not upgraded.
Need to get 0 B/90.8 MB of archives.
After this operation, 495 MB of additional disk space will be used.
Do you want to continue? [Y/n]  y
(Reading database ... 400759 files and directories currently installed.)
Preparing to unpack .../linux-image-6.13.6-surface-1_6.13.6-surface-1_amd64.deb 
...
Unpacking linux-image-6.13.6-surface-1 (6.13.6-surface-1) ...
Selecting previously unselected package linux-image-surface.
Preparing to unpack .../linux-image-surface_6.13.6-surface-1_amd64.deb ...
Unpacking linux-image-surface (6.13.6-surface-1) ...
Setting up linux-image-6.13.6-surface-1 (6.13.6-surface-1) ...
dkms: running auto installation service for kernel 6.13.6-surface-1.
Sign command: /lib/modules/6.13.6-surface-1/build/scripts/sign-file
Signing key: /var/lib/dkms/mok.key
Public certificate (MOK): /var/lib/dkms/mok.pub

Building module:
Cleaning build area...
make -j8 KERNELRELEASE=6.13.6-surface-1...(bad exit status: 2)
Error! Bad return status for module build on kernel: 6.13.6-surface-1 (x86_64)
Consult /var/lib/dkms/protection-agent/03.03.00.0050-c322816-h584516d6/build/mak
e.log for more information.
Sign command: /lib/modules/6.13.6-surface-1/build/scripts/sign-file
Signing key: /var/lib/dkms/mok.key
Public certificate (MOK): /var/lib/dkms/mok.pub

Building module:
Cleaning build area...
make -j8 KERNELRELEASE=6.13.6-surface-1 KERNEL_DIR=/lib/modules/6.13.6-surface-1
/build all...(bad exit status: 2)
Error! Bad return status for module build on kernel: 6.13.6-surface-1 (x86_64)
Consult /var/lib/dkms/v4l2loopback/0.12.7/build/make.log for more information.
Error! One or more modules failed to install during autoinstall.
Refer to previous errors for more information.
dkms: autoinstall for kernel: 6.13.6-surface-1 failed!
run-parts: /etc/kernel/postinst.d/dkms exited with return code 11
update-initramfs: Generating /boot/initrd.img-6.13.6-surface-1
Generating grub configuration file ...
Found background image: /usr/share/images/desktop-base/desktop-grub.png
Found linux image: /boot/vmlinuz-6.13.6-surface-1
Found initrd image: /boot/initrd.img-6.13.6-surface-1
Found linux image: /boot/vmlinuz-6.1.0-32-amd64
Found initrd image: /boot/initrd.img-6.1.0-32-amd64
Found linux image: /boot/vmlinuz-5.4.48-xe-max
Warning: os-prober will not be executed to detect other bootable partitions.
Systems on them will not be added to the GRUB boot configuration.
Check GRUB_DISABLE_OS_PROBER documentation entry.
Adding boot menu entry for UEFI Firmware Settings ...
done
dpkg: error processing package linux-image-6.13.6-surface-1 (--configure):
 installed linux-image-6.13.6-surface-1 package post-installation script subproc
ess returned error exit status 1
dpkg: dependency problems prevent configuration of linux-image-surface:
 linux-image-surface depends on linux-image-6.13.6-surface-1 (= 6.13.6-surface-1
); however:
  Package linux-image-6.13.6-surface-1 is not configured yet.

dpkg: error processing package linux-image-surface (--configure):
 dependency problems - leaving unconfigured
Errors were encountered while processing:
 linux-image-6.13.6-surface-1
 linux-image-surface
E: Sub-process /usr/bin/dpkg returned an error code (1
______________________________________________________________________________________________________________________________________
dkms log
//$ cat /var/lib/dkms/v4l2loopback/0.12.7/build/make.log
DKMS make.log for v4l2loopback-0.12.7 for kernel 6.13.6-surface-1 (x86_64)
Fri Apr  4 10:53:33 PM CEST 2025
Building v4l2-loopback driver...
make -C /lib/modules/6.13.6-surface-1/build M=/var/lib/dkms/v4l2loopback/0.12.7/build modules
make[1]: Entering directory '/usr/src/linux-headers-6.13.6-surface-1'
make[2]: Entering directory '/var/lib/dkms/v4l2loopback/0.12.7/build'
warning: the compiler differs from the one used to build the kernel
  The kernel was built by: gcc (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
  You are using:           gcc (Debian 12.2.0-14) 12.2.0
  CC [M]  v4l2loopback.o
v4l2loopback.c: In function ‘vidioc_querycap’:
v4l2loopback.c:717:9: error: implicit declaration of function ‘strlcpy’; did you mean ‘strncpy’? [-Werror=implicit-function-declaration]
  717 |         strlcpy(cap->driver, "v4l2 loopback", sizeof(cap->driver));
      |         ^~~~~~~
      |         strncpy
cc1: some warnings being treated as errors
make[4]: *** [/usr/src/linux-headers-6.13.6-surface-1/scripts/Makefile.build:196: v4l2loopback.o] Error 1
make[3]: *** [/usr/src/linux-headers-6.13.6-surface-1/Makefile:1991: .] Error 2
make[2]: *** [/usr/src/linux-headers-6.13.6-surface-1/Makefile:251: __sub-make] Error 2
make[2]: Leaving directory '/var/lib/dkms/v4l2loopback/0.12.7/build'
make[1]: *** [Makefile:251: __sub-make] Error 2
make[1]: Leaving directory '/usr/src/linux-headers-6.13.6-surface-1'
make: *** [Makefile:43: v4l2loopback.ko] Error 2
